controllers/extendeddaemonset/controller.go

```diff
@@ -425,6 +411,77 @@ func (r *Reconciler) selectNodes(logger logr.Logger, daemonsetSpec *datadoghqv1a
 	return nil
 }
 
+func isCanaryActive(daemonset *datadoghqv1alpha1.ExtendedDaemonSet, activeERSName string, uptodateERSName string, isCanaryFailed bool) bool {
+	if daemonset.Spec.Strategy.Canary == nil {
+		return false
+	}
+
+	if isCanaryFailed || activeERSName == uptodateERSName {
+		return false
+	}
+
+	return true
+}
+
+func manageCanaryStatusConditions(status *datadoghqv1alpha1.ExtendedDaemonSetStatus, now metav1.Time, isCanaryFailed bool, isCanaryPaused bool, pausedReason datadoghqv1alpha1.ExtendedDaemonSetStatusReason, ersName string) *datadoghqv1alpha1.ExtendedDaemonSetStatus {
+	updateOptions := &conditions.UpdateConditionOptions{
+		IgnoreFalseConditionIfNotExist: false,
+		SupportLastUpdate:              false,
+	}
+	if isCanaryFailed {
+		msg := fmt.Sprintf("canary failed with ers: %s", ersName)
+		conditions.UpdateExtendedDaemonSetStatusCondition(status, now, datadoghqv1alpha1.ConditionTypeEDSCanaryFailed, corev1.ConditionTrue, "CanaryFailed", msg, updateOptions)
+		status.State = datadoghqv1alpha1.ExtendedDaemonSetStatusStateCanaryFailed
+	} else {
+		conditions.UpdateExtendedDaemonSetStatusCondition(status, now, datadoghqv1alpha1.ConditionTypeEDSCanaryFailed, corev1.ConditionFalse, "", "", updateOptions)
+	}
+
+	if isCanaryPaused && !isCanaryFailed {
+		msg := fmt.Sprintf("canary paused with ers: %s", ersName)
+		conditions.UpdateExtendedDaemonSetStatusCondition(status, now, datadoghqv1alpha1.ConditionTypeEDSCanaryPaused, corev1.ConditionTrue, string(pausedReason), msg, updateOptions)
+		status.State = datadoghqv1alpha1.ExtendedDaemonSetStatusStateCanaryPaused
+		status.Reason = pausedReason
+	} else {
+		conditions.UpdateExtendedDaemonSetStatusCondition(status, now, datadoghqv1alpha1.ConditionTypeEDSCanaryPaused, corev1.ConditionFalse, "", "", updateOptions)
+	}
+	return status
+}
+
+func manageCanaryStatus(status *datadoghqv1alpha1.ExtendedDaemonSetStatus, upToDate *datadoghqv1alpha1.ExtendedDaemonSetReplicaSet, isCanaryActive bool, isCanaryFailed bool, isCanaryPaused bool) *datadoghqv1alpha1.ExtendedDaemonSetStatus {
+	switch {
+	case isCanaryFailed:
+		// Canary deployment is no longer needed because it was marked as failed
+		// This `if` block needs to be first to respect Failed Canary state until annotation is removed
+		status.Canary = nil
+		status.State = datadoghqv1alpha1.ExtendedDaemonSetStatusStateCanaryFailed
+		status.Reason = ""
```

Comment:
do you think it's worth it to put the failed reason here?

Reply:
no really because it will be cleanup right after, when the controller will consider currentERS as active in the next reconcile loop.

But, to know the fail reason, we can now look at the failed condition that will not be removed. (only step to false, we the EDS will be back to currentERS)